packages/deployments/contracts/contracts/nomad-xapps/contracts/promise-router/PromiseRouter.sol

```diff
@@ -154,26 +187,58 @@ contract PromiseRouter is Version0, Router {
 
     bytes32 transferId = _msg.transferId();
     address callbackAddress = _msg.callbackAddress();
-    bytes memory data = _msg.returnCallData();
-    
-    //TODO process callback
+    bool success = _msg.returnSuccess();
+    bytes memory data = _msg.returnData();
+
+    //store Promise message
+    promiseMessages[transferId] = _message;
 
     // emit Receive event
-    emit Receive(_originAndNonce(_origin, _nonce), _origin, transferId, callbackAddress, data);
+    emit Receive(_originAndNonce(_origin, _nonce), _origin, transferId, callbackAddress, success, data, _message);
   }
 
-  function process(bytes32 transactionId, bytes memory _message) internal {
-    // Should parse out the return data and callback address from message
+  function process(bytes32 transferId) public payable nonReentrant {
+    // parse out the return data and callback address from message
+    bytes memory _message = promiseMessages[transferId];
+    if (_message.length == 0) revert PromiseRouter__process_InvalidMessage();
 
-    // Should execute the callback() function on the provided Callback address
+    bytes29 _msg = _message.ref(0).mustBePromiseCallback();
 
-    // Should enforce relayer is whitelisted by calling local connext contract
+    // enforce relayer is whitelisted by calling local connext contract
+    if (!connext.isApprovedRelayer(msg.sender)) revert PromiseRouter__process_notApprovedRelayer();
 
     // Should transfer the stored relayer fee to the msg.sender
+    // TODO check if callbackFee which already paid is sufficient or not
+    //      If not, should revert and should add more fee.
+    // if (gasFee > callbackFees[transferId]) {
+    //   revert PromiseRouter__process_insufficientCallbackFee();
+    // }
```

Comment:
I think we can remove this `TODO` -- we should have users in charge of setting their own fee rather than relying on the domain to establish it

Reply:
Okay